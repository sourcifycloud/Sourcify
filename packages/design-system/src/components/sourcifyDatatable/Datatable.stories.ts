import sourcifyDatatable from './Datatable.vue';
import type { StoryFn } from '@storybook/vue';
import { rows, columns } from './__tests__/data';

export default {
	title: 'Atoms/Datatable',
	component: sourcifyDatatable,
};

export const Default: StoryFn = (args, { argTypes }) => ({
	props: Object.keys(argTypes),
	components: {
		sourcifyDatatable,
	},
	template: '<sourcify-datatable v-bind="$props"></sourcify-datatable>',
});

Default.args = {
	columns,
	rows,
};
