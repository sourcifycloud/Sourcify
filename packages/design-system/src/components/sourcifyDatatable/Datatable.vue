<script lang="ts">
import { computed, defineComponent, PropType, ref, useCssModule } from 'vue';
import { DatatableColumn, DatatableRow, DatatableRowDataType } from '../../types';
import { getValueByPath } from '../../utils';
import { useI18n } from '../../composables';
import sourcifySelect from '../sourcifySelect';
import sourcifyOption from '../sourcifyOption';
import sourcifyPagination from '../sourcifyPagination';

export default defineComponent({
	name: 'sourcify-datatable',
	components: {
		sourcifySelect,
		sourcifyOption,
		sourcifyPagination,
	},
	emits: ['update:currentPage', 'update:rowsPerPage'],
	props: {
		columns: {
			type: Array as PropType<DatatableColumn[]>,
			required: true,
		},
		rows: {
			type: Array as PropType<DatatableRow[]>,
			required: true,
		},
		currentPage: {
			type: Number,
			default: 1,
		},
		pagination: {
			type: Boolean,
			default: true,
		},
		rowsPerPage: {
			type: [Number, String] as PropType<number | '*'>,
			default: 10,
		},
	},
	setup(props, { emit }) {
		const { t } = useI18n();
		const rowsPerPageOptions = ref([10, 25, 50, 100]);

		const style = useCssModule();

		const totalPages = computed(() => {
			if (props.rowsPerPage === '*') {
				return 1;
			}

			return Math.ceil(props.rows.length / props.rowsPerPage);
		});

		const totalRows = computed(() => {
			return props.rows.length;
		});

		const visibleRows = computed(() => {
			if (props.rowsPerPage === '*') {
				return props.rows;
			}

			const start = (props.currentPage - 1) * props.rowsPerPage;
			const end = start + props.rowsPerPage;

			return props.rows.slice(start, end);
		});

		const classes = computed(() => {
			return {
				datatable: true,
				[style.datatableWrapper]: true,
			};
		});

		function onUpdateCurrentPage(value: number) {
			emit('update:currentPage', value);
		}

		function onRowsPerPageChange(value: number | '*') {
			emit('update:rowsPerPage', value);

			const maxPage = value === '*' ? 1 : Math.ceil(totalRows.value / value);
			if (maxPage < props.currentPage) {
				onUpdateCurrentPage(maxPage);
			}
		}

		function getTdValue(row: DatatableRow, column: DatatableColumn) {
			return getValueByPath<DatatableRowDataType>(row, column.path);
		}

		function getThStyle(column: DatatableColumn) {
			return {
				...(column.width ? { width: column.width } : {}),
			};
		}

		return {
			t,
			classes,
			totalPages,
			totalRows,
			visibleRows,
			rowsPerPageOptions,
			getTdValue,
			getThStyle,
			onUpdateCurrentPage,
			onRowsPerPageChange,
		};
	},
});
</script>

<template>
	<div :class="classes" v-on="$listeners">
		<table :class="$style.datatable">
			<thead :class="$style.datatableHeader">
				<tr>
					<th
						v-for="column in columns"
						:key="column.id"
						:class="column.classes"
						:style="getThStyle(column)"
					>
						{{ column.label }}
					</th>
				</tr>
			</thead>
			<tbody>
				<template v-for="row in visibleRows">
					<slot name="row" :columns="columns" :row="row" :getTdValue="getTdValue">
						<tr :key="row.id">
							<td v-for="column in columns" :key="column.id" :class="column.classes">
								<component v-if="column.render" :is="column.render" :row="row" :column="column" />
								<span v-else>{{ getTdValue(row, column) }}</span>
							</td>
						</tr>
					</slot>
				</template>
			</tbody>
		</table>

		<div :class="$style.pagination">
			<sourcify-pagination
				v-if="totalPages > 1"
				background
				:pager-count="5"
				:page-size="rowsPerPage"
				layout="prev, pager, next"
				:total="totalRows"
				:currentPage="currentPage"
				@update:currentPage="onUpdateCurrentPage"
			/>

			<div :class="$style.pageSizeSelector">
				<sourcify-select
					size="mini"
					:value="rowsPerPage"
					@input="onRowsPerPageChange"
					popper-append-to-body
				>
					<template #prepend>{{ t('datatable.pageSize') }}</template>
					<sourcify-option
						v-for="size in rowsPerPageOptions"
						:key="size"
						:label="`${size}`"
						:value="size"
					/>
					<sourcify-option :label="`All`" value="*"> </sourcify-option>
				</sourcify-select>
			</div>
		</div>
	</div>
</template>

<style lang="scss" module>
.datatableWrapper {
	display: block;
	width: 100%;
}

.datatable {
	width: 100%;

	tbody {
		tr {
			td {
				vertical-align: top;
				color: var(--color-text-base);
				padding: var(--spacing-s) var(--spacing-2xs);
			}

			&:nth-of-type(even) {
				background: var(--color-background-xlight);
			}

			&:nth-of-type(odd) {
				background: var(--color-background-light);
			}
		}
	}
}

.datatableHeader {
	background: var(--color-background-base);

	th {
		text-align: left;
		padding: var(--spacing-s) var(--spacing-2xs);
	}
}

.pagination {
	width: 100%;
	display: flex;
	justify-content: center;
	align-items: center;
	bottom: 0;
	overflow: visible;
	margin-top: var(--spacing-s);
}

.pageSizeSelector {
	text-transform: capitalize;
	max-width: 150px;
	flex: 0 1 auto;
}
</style>
