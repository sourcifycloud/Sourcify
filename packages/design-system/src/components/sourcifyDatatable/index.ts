import sourcifyDatatable from './Datatable.vue';

export default sourcifyDatatable;
