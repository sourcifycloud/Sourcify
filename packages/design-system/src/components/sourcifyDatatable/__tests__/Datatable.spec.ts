import { render } from '@testing-library/vue';
import sourcifyDatatable from '../Datatable.vue';
import { rows, columns } from './data';

const stubs = ['sourcify-select', 'sourcify-option', 'sourcify-button', 'sourcify-pagination'];

describe('components', () => {
	describe('sourcifyDatatable', () => {
		const rowsPerPage = 10;

		it('should render correctly', () => {
			const wrapper = render(sourcifyDatatable, {
				propsData: {
					columns,
					rows,
					rowsPerPage,
				},
				stubs,
			});

			expect(wrapper.container.querySelectorAll('thead tr').length).toEqual(1);
			expect(wrapper.container.querySelectorAll('tbody tr').length).toEqual(rowsPerPage);
			expect(wrapper.container.querySelectorAll('tbody tr td').length).toEqual(
				columns.length * rowsPerPage,
			);
			expect(wrapper.html()).toMatchSnapshot();
		});

		it('should add column classes', () => {
			const wrapper = render(sourcifyDatatable, {
				propsData: {
					columns: columns.map((column) => ({ ...column, classes: ['example'] })),
					rows,
					rowsPerPage,
				},
				stubs,
			});

			expect(wrapper.container.querySelectorAll('.example').length).toEqual(
				columns.length * (rowsPerPage + 1),
			);
		});

		it('should render row slot', () => {
			const wrapper = render(sourcifyDatatable, {
				propsData: {
					columns,
					rows,
					rowsPerPage,
				},
				stubs,
				scopedSlots: {
					row: '<main><td v-for="column in props.columns" :key="column.id">Row slot</td></main>', // Wrapper is necessary for looping
				},
			});

			expect(wrapper.container.querySelectorAll('tbody td').length).toEqual(
				columns.length * rowsPerPage,
			);
			expect(wrapper.container.querySelector('tbody td')?.textContent).toEqual('Row slot');
		});
	});
});
