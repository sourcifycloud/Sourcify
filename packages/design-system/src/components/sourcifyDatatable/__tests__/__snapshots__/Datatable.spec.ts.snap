// Vitest Snapshot v1

exports[`components > sourcifyDatatable > should render correctly 1`] = `
"<div class=\\"datatable datatableWrapper\\">
  <table class=\\"datatable\\">
    <thead class=\\"datatableHeader\\">
      <tr>
        <th> ID </th>
        <th> Name </th>
        <th> Age </th>
        <th> Action </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td><span>1</span></td>
        <td><span>Richard Hendricks</span></td>
        <td><span>29</span></td>
        <td><button aria-live=\\"polite\\" class=\\"button button primary medium\\" column=\\"[object Object]\\">
            <!----><span>Button 1</span></button></td>
      </tr>
      <tr>
        <td><span>2</span></td>
        <td><span>Bertram Gilfoyle</span></td>
        <td><span>44</span></td>
        <td><button aria-live=\\"polite\\" class=\\"button button primary medium\\" column=\\"[object Object]\\">
            <!----><span>Button 2</span></button></td>
      </tr>
      <tr>
        <td><span>3</span></td>
        <td><span>Dinesh Chugtai</span></td>
        <td><span>31</span></td>
        <td><button aria-live=\\"polite\\" class=\\"button button primary medium\\" column=\\"[object Object]\\">
            <!----><span>Button 3</span></button></td>
      </tr>
      <tr>
        <td><span>4</span></td>
        <td><span>Jared Dunn </span></td>
        <td><span>38</span></td>
        <td><button aria-live=\\"polite\\" class=\\"button button primary medium\\" column=\\"[object Object]\\">
            <!----><span>Button 4</span></button></td>
      </tr>
      <tr>
        <td><span>5</span></td>
        <td><span>Richard Hendricks</span></td>
        <td><span>29</span></td>
        <td><button aria-live=\\"polite\\" class=\\"button button primary medium\\" column=\\"[object Object]\\">
            <!----><span>Button 5</span></button></td>
      </tr>
      <tr>
        <td><span>6</span></td>
        <td><span>Bertram Gilfoyle</span></td>
        <td><span>44</span></td>
        <td><button aria-live=\\"polite\\" class=\\"button button primary medium\\" column=\\"[object Object]\\">
            <!----><span>Button 6</span></button></td>
      </tr>
      <tr>
        <td><span>7</span></td>
        <td><span>Dinesh Chugtai</span></td>
        <td><span>31</span></td>
        <td><button aria-live=\\"polite\\" class=\\"button button primary medium\\" column=\\"[object Object]\\">
            <!----><span>Button 7</span></button></td>
      </tr>
      <tr>
        <td><span>8</span></td>
        <td><span>Jared Dunn </span></td>
        <td><span>38</span></td>
        <td><button aria-live=\\"polite\\" class=\\"button button primary medium\\" column=\\"[object Object]\\">
            <!----><span>Button 8</span></button></td>
      </tr>
      <tr>
        <td><span>9</span></td>
        <td><span>Richard Hendricks</span></td>
        <td><span>29</span></td>
        <td><button aria-live=\\"polite\\" class=\\"button button primary medium\\" column=\\"[object Object]\\">
            <!----><span>Button 9</span></button></td>
      </tr>
      <tr>
        <td><span>10</span></td>
        <td><span>Bertram Gilfoyle</span></td>
        <td><span>44</span></td>
        <td><button aria-live=\\"polite\\" class=\\"button button primary medium\\" column=\\"[object Object]\\">
            <!----><span>Button 10</span></button></td>
      </tr>
    </tbody>
  </table>
  <div class=\\"pagination\\">
    <sourcify-pagination-stub pagesize=\\"10\\" total=\\"15\\" pagercount=\\"5\\" currentpage=\\"1\\" layout=\\"prev, pager, next\\" pagesizes=\\"10,20,30,40,50,100\\" background=\\"true\\"></sourcify-pagination-stub>
    <div class=\\"pageSizeSelector\\">
      <sourcify-select-stub value=\\"10\\" size=\\"mini\\" popperappendtobody=\\"true\\">
        <sourcify-option-stub value=\\"10\\" label=\\"10\\"></sourcify-option-stub>
        <sourcify-option-stub value=\\"25\\" label=\\"25\\"></sourcify-option-stub>
        <sourcify-option-stub value=\\"50\\" label=\\"50\\"></sourcify-option-stub>
        <sourcify-option-stub value=\\"100\\" label=\\"100\\"></sourcify-option-stub>
        <sourcify-option-stub value=\\"*\\" label=\\"All\\"></sourcify-option-stub>
      </sourcify-select-stub>
    </div>
  </div>
</div>"
`;
